distributedsocialnetwork/author/retrieval.py

```diff
@@ -0,0 +1,87 @@
+from post.retrieval import sanitize_author, sanitize_post, transformSource
+from friend.models import Friend
+from author.serializers import AuthorSerializer
+import requests
+import datetime
+import uuid
+from django.conf import settings
+from django.shortcuts import get_object_or_404
+
+# Given a QuerySet 'posts', returns another QuerySet
+# containing the posts that 'author_id' can see.
+def filter_posts(posts,author_id):
+    public_posts = posts.filter(visibility="PUBLIC")
+    user_posts = posts.filter(author=author_id)
+    privated_posts = posts.filter(
+        visibility="PRIVATE", visibleTo__icontains=author_id)
+    serveronly_posts = posts.filter(
+        visibility="SERVERONLY", author__in=Friend.objects.get_friends(author_id))
+    friend_posts = posts.filter(
+        visibility="FRIENDS", author__in=Friend.objects.get_friends(author_id))
+    foaf_posts = posts.filter(
+        visibility="FOAF", author__in=Friend.objects.get_foaf(author_id))
+    post_query_set = public_posts | user_posts | privated_posts | serveronly_posts | friend_posts | foaf_posts
+    return post_query_set
+
+# Returns a QuerySet of all foreign posts the specified author is authorized to see
+def get_visible_posts(author_id):
+    author = get_object_or_404(Author, id=author_id)
+    nodes = list(Node.objects.all())
+    visible_posts = Post.objects.none()
+    for node in nodes:
+        if node.node_auth_username != "":
+            # We have a set username and password to authenticate with this node.
+            # We send a GET request to their endpoint.
+            url = node.api_url + 'author/posts'
+            print("sending to ", url)
+            # Include a header containing the Author's id, other nodes may not conform to this.
+            response = requests.get(
+                url, auth=(node.node_auth_username, node.node_auth_password), headers={'content-type': 'application/json', 'Accept': 'application/json', 'AuthorId': author.id })
+            if response.status_code == 200:
+                # print(response.json())
+                # We have the geen light to continue. Otherwise, we just use what we have cached.
+                posts_json = response.json()
+                for post in posts_json["posts"]:
+                    # We first have to ensure the author of each post is in our database.
+
+                    author = sanitize_author(post["author"])
+                    post = sanitize_post(post)
+                    post = transformSource(post)
+                    author['displayName'] = author['displayName'] + \
+                        " (" + node.server_username + ")"
+                    author_parts = author['id'].split('/')
+                    authorID = author_parts[-1]
+                    if authorID == '':
+                        authorID = author_parts[-2]
+                    author['url'] = settings.FORMATTED_HOST_NAME + \
+                        'author/' + authorID
```

Comment:
I would recommend adding this in here:
```
if (len(Author.objects.filter(id=author['id'])) == 1):
                        old_author = Author.objects.get(id=author['id'])
                        author_serializer = AuthorSerializer(
                            old_author, data=author)
                    else:
                        author_serializer = AuthorSerializer(data=author)
```

This just ensures that, if we already have an entry for the author in our database, we are updating them instead of trying to insert a new one (which would fail).

Reply:
Good call, thanks!